### PR TITLE
More javadocs, refactoring, and unit tests

### PR DESCRIPTION
**Pull request by david-waltermire**
# Committer Notes

Added Javadocs and performed light refactoring.

Focused quite a bit on improving unit test coverage over Metapath operations, specifically the math-related operations.

### All Submissions:

- [x] Have you selected the correct base branch per [Contributing](https://github.com/metaschema-framework/metaschema-java/blob/main/CONTRIBUTING.md) guidance?
- [x] Have you followed the guidelines in our [Contributing](https://github.com/metaschema-framework/metaschema-java/blob/main/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/metaschema-framework/metaschema-java/pulls) for the same update/change?
- [ ] Have you squashed any non-relevant commits and commit messages? \[[instructions](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History)\]
- [x] Have you set "[Allow edits and access to secrets by maintainers](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork)"?

By submitting a pull request, you are agreeing to provide this contribution under the [CC0 1.0 Universal public domain](https://creativecommons.org/publicdomain/zero/1.0/) dedication.

### Changes to Core Features:

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Have you written new tests for your core changes, as applicable?
- [ ] Have you included examples of how to use your new feature(s)?
- [ ] Have you updated all website and readme documentation affected by the changes you made?

## Release Notes

- **New Features**
  - Added support for TIME data type.
  - Enhanced function and context handling in Metapath expressions.
  - Introduced new utility methods for string and collection manipulation.
  - Added new methods for addition and subtraction of durations to time items.
  - Introduced arithmetic operations for numeric items, including addition, subtraction, multiplication, and division.
  - Added new methods for arithmetic operations on decimal items.

- **Improvements**
  - Refactored sequence and collection handling for better performance.
  - Enhanced error handling and validation mechanisms.
  - Improved documentation across multiple classes and interfaces.
  - Updated caching and lookup mechanisms for qualified names.
  - Enhanced clarity in exception handling and validation logic.
  - Streamlined the conversion process for integer items in various methods.
  - Improved handling of arithmetic operations and overflow checks.

- **Bug Fixes**
  - Fixed various edge cases in arithmetic and comparison operations.
  - Improved handling of sequences and iterables.
  - Resolved potential null pointer and type casting issues.

- **Breaking Changes**
  - Renamed and reorganized several classes and methods.
  - Modified method signatures in core expression and utility classes.
  - Updated exception handling in multiple components.

- **Testing**
  - Added comprehensive test cases for arithmetic and path expressions.
  - Improved test coverage for utility methods and edge cases.
  - Transitioned to Mockito for mocking dependencies in tests.
  - Updated tests to reflect changes in method signatures and exception handling.

### CODE REVIEW

**Comment by david-waltermire** [comment]
@coderabbitai full review

File: core/src/main/java/gov/nist/secauto/metaschema/core/util/StringUtils.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+   */
+  @NonNull
+  @SuppressWarnings("null")
+  @SuppressFBWarnings("NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE")
+  public static String requireNonEmpty(@Nullable String string, @NonNull String message) {
+    if (string.isEmpty()) {
+      throw new IllegalArgumentException(message);
+    }
+    return string;
+  }
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/util/StringUtils.java:67]: The method is designed to throw a NPE if the provided string is null.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/cst/logic/AbstractFilterExpression.java [hunk 2]
```diff
@@ -53,9 +53,7 @@ public ISequence<? extends IItem> accept(
 
     ISequence<?> left = getLeft().accept(dynamicContext, focus);
     ISequence<?> right = getRight().accept(dynamicContext, focus);
-    List<? extends IItem> rightList = right.getValue();
-
-    return applyFilterTo(left, rightList);
+    return applyFilterTo(left, right);
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/cst/logic/AbstractFilterExpression.java:56]: An ISequence implements the list interface, so this is ok. Adjusted the applyFilterTo Javadoc below.

File: core/src/main/java/gov/nist/secauto/metaschema/core/util/StringUtils.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+   */
+  @NonNull
+  @SuppressWarnings("null")
+  @SuppressFBWarnings("NP_PARAMETER_MUST_BE_NONNULL_BUT_MARKED_AS_NULLABLE")
+  public static String requireNonEmpty(@Nullable String string) {
+    if (string.isEmpty()) {
+      throw new IllegalArgumentException("String is empty.");
+    }
+    return string;
+  }
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/util/StringUtils.java]: This was addressed by ensuring that the provided string argument is `@NonNull`.

File: core/src/test/java/gov/nist/secauto/metaschema/core/metapath/item/node/MockNodeItemFactory.java [hunk 4]
```diff
@@ -159,10 +216,21 @@ public IFieldNodeItem field(
...
+   * Construct a new mocked field node item with no child flags.
+   *
+   * @param name
+   *          the qualified name of the field node item
+   * @param flags
+   *          the node item's child flag node items
+   * @param modelItems
+   *          the node item's child model node items
+   * @return the mocked field node item
+   */
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply core/src/test/java/gov/nist/secauto/metaschema/core/metapath/item/node/MockNodeItemFactory.java]: I updated the Javadoc accordingly.

**Comment by david-waltermire** [comment]
@coderabbitai I pushed fixes that address your prior feedback in commits 8407f8f, 7072a0f, and 7b1fd05. Please review these commits.

**Comment by david-waltermire** [comment]
@coderabbitai summary

Please regenerating the summary for this PR , incorporating your analysis from your last discussion message and any additional insights from the recent changes.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/cst/AbstractCSTVisitorBase.java [hunk 5]
```diff
@@ -28,6 +28,24 @@
...
+ * Provides utility methods for processing Metapath abstract syntax tree (AST)
+ * nodes to produce a compact syntax tree (CST).
+ * <p>
+ * This base class implements common visitor patterns for transforming AST nodes
+ * into a more compact representation. The CST is optimized for efficient
+ * evaluation of Metapath expressions.
+ * <p>
+ * Key utility methods include:
+ * <ul>
+ * <li>{@link #nairyToList} - Processes n-ary expressions into a list</li>
```
Comment (aj-stein-gsa) [inline core/src/main/java/gov/nist/secauto/metaschema/core/metapath/cst/AbstractCSTVisitorBase.java:41]: Potentially blocking because of function name, not just docs: sorry to be "that person," but I noticed this before and avoided bringing it up out of context. We do mean "[n-ary](https://xlinux.nist.gov/dads/HTML/naryfunc.html)" not n-airy, right?

Do we want to fix this in docs and functions?

Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/cst/AbstractCSTVisitorBase.java]: I fixed this in the documentation. Thanks!

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/impl/OperationFunctions.java [hunk 6]
```diff
@@ -290,8 +361,7 @@
     Duration duration1 = arg1.asDuration();
     Duration duration2 = arg2.asDuration();
 
-    return IDayTimeDurationItem.valueOf(
-        ObjectUtils.notNull(duration1.minus(duration2)));
+    return IDayTimeDurationItem.valueOf(ObjectUtils.notNull(duration2.minus(duration1)));
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/impl/OperationFunctions.java]: This has been fixed.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/impl/OperationFunctions.java [hunk 7]
```diff
@@ -268,9 +341,7 @@ public static IYearMonthDurationItem opSubtractYearMonthDurations(
       @NonNull IYearMonthDurationItem arg2) {
     Period duration1 = arg1.asPeriod();
     Period duration2 = arg2.asPeriod();
-
-    return IYearMonthDurationItem.valueOf(
-        ObjectUtils.notNull(duration1.minus(duration2)));
+    return IYearMonthDurationItem.valueOf(ObjectUtils.notNull(duration2.minus(duration1)));
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/impl/OperationFunctions.java]: This has been fixed.

File: core/src/main/java/gov/nist/secauto/metaschema/core/util/StringUtils.java [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+   *          the pattern to search for
+   * @param replacementFunction
+   *          a function that will provided the replacement text
+   * @return the resulting text after replacing matching occurrences in
+   *         {@code text}
+   */
+  public static CharSequence replaceTokens(
+      @NonNull CharSequence text,
+      @NonNull Pattern pattern,
+      Function<Matcher, CharSequence> replacementFunction) {
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/util/StringUtils.java]: This has been fixed.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/impl/AbstractSequence.java [hunk 9]
```diff
@@ -41,12 +45,171 @@ public static <T extends IItem> ISequence<T> empty() {
...
+      return false;
+    }
+    Iterator<?> iter = ((List<?>) other).iterator();
+    for (Object element : asList()) {
+      if (!iter.hasNext() || !element.equals(iter.next())) {
+        return false;
+      }
+    }
+    return !iter.hasNext();
+  }
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/impl/AbstractSequence.java]: This has been fixed.

**Comment by david-waltermire** [comment]
@aj-stein-gsa I am going to work on adding more unit tests for the math related functions and expressions, before calling this done. I'll ping you when this is ready for re-review.

**Comment by david-waltermire** [comment]
@coderabbitai review

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/ArrayRemove.java [hunk 10]
```diff
@@ -93,7 +93,7 @@ public static <T extends IItem> IArrayItem<T> removeItems(
     return remove(
         array,
         ObjectUtils.notNull(positions.stream()
-            .map(position -> position.asInteger().intValueExact())
+            .map(IIntegerItem::toIntValueExact)
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/library/ArrayRemove.java:97]: This was fixed.

File: core/src/test/java/gov/nist/secauto/metaschema/core/metapath/function/impl/OperationFunctionsTest.java [hunk 11]
```diff
@@ -0,0 +1,452 @@
...
+        @NonNull INumericItem dividend,
+        @NonNull INumericItem divisor) {
+      ArithmeticFunctionException thrown = assertThrows(ArithmeticFunctionException.class, () -> {
+        OperationFunctions.opNumericDivide(dividend, divisor);
+      });
+      assertEquals(ArithmeticFunctionException.DIVISION_BY_ZERO, thrown.getCode());
+    }
+
+    private Stream<Arguments> provideValuesOpNumericIntegerDivide() {
+      Arguments.of(
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply core/src/test/java/gov/nist/secauto/metaschema/core/metapath/function/impl/OperationFunctionsTest.java]: This was fixed.

**Comment by david-waltermire** [comment]
@coderabbitai review

Please update the PR release notes.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/impl/OperationFunctions.java [hunk 12]
```diff
@@ -826,17 +947,7 @@ public static INumericItem opNumericMultiply(@NonNull INumericItem left, @NonNul
...
-
-    if (BigDecimal.ZERO.equals(decimalDivisor)) {
-      throw new ArithmeticFunctionException(ArithmeticFunctionException.DIVISION_BY_ZERO,
-          ArithmeticFunctionException.DIVISION_BY_ZERO_MESSAGE);
-    }
-
-    BigDecimal decimalDividend = dividend.asDecimal();
-    return IDecimalItem.valueOf(
-        ObjectUtils.notNull(decimalDividend.divide(decimalDivisor, FunctionUtils.MATH_CONTEXT)));
+    return ((IDecimalItem) dividend).divide((IDecimalItem) divisor);
```
[the remark this answers is not in the document]
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/function/impl/OperationFunctions.java:950]: I'll add a throws javadoc to address this, since the called method already checks this.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/atomic/INumericItem.java [hunk 13]
```diff
@@ -78,6 +79,19 @@ static INumericItem cast(@NonNull IAnyAtomicItem item) {
...
+    try {
+      return asInteger().intValueExact();
+    } catch (ArithmeticException ex) {
+      throw new CastFunctionException(
+          CastFunctionException.INPUT_VALUE_TOO_LARGE,
+          this,
+          String.format("Input value '%s' is out of range for a Java int.", asString()),
+          ex);
+    }
+  }
```
Comment (david-waltermire) [inline core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/atomic/INumericItem.java:93]: @coderabbitai Please propose specialized implementations for integer and decimal types to address the FIXME comment

Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/atomic/INumericItem.java:93]: Thanks. I'll integrate these into the code, but I am going to keep the method in INumericItem without a default body, since the method signature is consistent across the implementations.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/ICollectionValue.java [hunk 14]
```diff
@@ -27,7 +27,6 @@ public interface ICollectionValue {
    *
    * @return the resulting sequence
    */
-  // TODO: rename to toSequence and resolve conflicting methods?
   @NonNull
   default ISequence<?> toSequence() {
     return this instanceof ISequence
```
Comment (aj-stein-gsa) [inline core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/ICollectionValue.java:32]: Is this really resolved? If so I find the function name very confusing still.

**Review by aj-stein-gsa — APPROVED** [review]
Looks good but I found one minor source of confusion re a function name. On the whole, solid work and the cleaned up util functions and comprehensive docs will help me.

File: core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/ICollectionValue.java (hunk 14 above)
Comment (david-waltermire) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/ICollectionValue.java:32]: This was addressed in a PR a while ago. This comment was never cleaned up.

Comment (aj-stein-gsa) [reply core/src/main/java/gov/nist/secauto/metaschema/core/metapath/item/ICollectionValue.java:32]: I missed part of the ternary function that was important while reviewing on the work phone even in horizontal view because the diffed comment and first return line is all I saw.

Apologies. 🤦
